### PR TITLE
Add property-based filtering (`-filter` keyword arg) on all SDC `get_*` commands (+ DRY refactor)

### PR DESCRIPTION
**Pull request by akashlevy**
Creating a draft PR for this. Basic goal is to add `-filter` to all SDC `get_*` commands to match the behavior of commercial STA tools. I'll be following DRY principle as much as possible in my commits.

Need some input: how to deal with error codes? There's no room to squeeze in new codes, so right now I'm adding new ones by adding 2000 to the previous error code. Not sure if you have a preferred way of adding new codes into the mix. Another option is to remove some codes and have the `filter_objs` function have a single error code no matter which object type is being filtered.

**Tasks:**
- [x] Add generic `filter_objs` TCL function
- [x] Remove global `filter_*` regexp variables and instead move these inside of `filter_objs`
- [x] Refactor `-filter` for `get_cells`
    - [x] Use `filter_objs`
    - [x] Delete `filter_insts1`
- [x] Add `-filter` to `get_clocks`
    - [x] TCL: Add `[-filter expr]` to header
    - [x] TCL: Add `-filter` to key flag parsing
    - [x] TCL: Add filter handling in body
    - [x] SWIG: Add `filter_clocks` function
    - [x] SWIG: Add relevant datatypes (in `ClockSeq *`)
- [x] Add `-filter` to `get_lib_cells`
    - [x] TCL: Add `[-filter expr]` to header
    - [x] TCL: Add `-filter` to key flag parsing
    - [x] TCL: Add filter handling in body
    - [x] SWIG: Add `filter_lib_cells` function
    - [x] SWIG: Add relevant datatypes (in `LibertyCellSeq *`)
- [x] Add `-filter` to `get_lib_pins`
    - [x] TCL: Add `[-filter expr]` to header
    - [x] TCL: Add `-filter` to key flag parsing
    - [x] TCL: Add filter handling in body
    - [x] SWIG: Add `filter_lib_cells` function
    - [x] SWIG: Add relevant datatypes (in `LibertyPortSeq *`)
- [x] Add `-filter` to `get_libs`
    - [x] TCL: Add `[-filter expr]` to header
    - [x] TCL: Add `-filter` to key flag parsing
    - [x] TCL: Add filter handling in body
    - [x] SWIG: Add `filter_liberty_libraries` function
    - [x] SWIG: Add relevant datatypes (out `LibertyLibrarySeq`, in `LibertyLibrarySeq *`)
- [x] Add `-filter` to `get_nets`
    - [x] TCL: Add `[-filter expr]` to header
    - [x] TCL: Add `-filter` to key flag parsing
    - [x] TCL: Add filter handling in body
    - [x] SWIG: Add `filter_nets` function
    - [x] SWIG: Add relevant datatypes (in `NetSeq *`)
- [x] Refactor `-filter` for `get_pins`
    - [x] Use `filter_objs`
    - [x] Delete `filter_pins1`
- [x] Refactor `-filter` for `get_ports`
    - [x] Use `filter_objs`
    - [x] Delete `filter_ports1`
- [x] Refactor `-filter` for `get_timing_edges`
    - [x] Use `filter_objs`
    - [x] Delete `filter_timing_arcs1`
- [x] Add some simple tests for `-filter` flag with `asap7` collateral
    - [x] `get_cells`
    - [x] `get_clocks`
    - [x] `get_lib_cells`
    - [x] `get_lib_pins`
    - [x] `get_libs`
    - [x] `get_nets`
    - [x] `get_pins`
    - [x] `get_ports`
- [x] Update docs to include newly-added `-filter` flags
    - [x] `get_clocks`
    - [x] `get_lib_cells`
    - [x] `get_lib_pins`
    - [x] `get_libs`
    - [x] `get_nets`
- [x] Update error code numbers
    - [x] Update `Sdc.tcl`
    - [x] Update `messages.txt`

### CODE REVIEW

**Comment by akashlevy** [comment]
All right, ready for review. Please let me know how to proceed regarding the error codes.

File: sdc/Sdc.tcl [hunk 1]
```diff
@@ -406,6 +406,36 @@ proc current_design { {design ""} } {
 
 ################################################################
 
+# Generic filter_objs command.
+proc filter_objs { filter objects filter_function filter_type error_code } {
+  set filter_regexp1 {@?([a-zA-Z_]+) *(==|!=|=~|!~) *([0-9a-zA-Z_\*]+)}
+  set filter_or_regexp "($filter_regexp1) *\\|\\| *($filter_regexp1)"
```
Comment (jjcherry56) [inline sdc/Sdc.tcl:412]: filter_or_regexp and filter_and_regexp are referenced here but the definitions were removed

File: sdc/Sdc.tcl [hunk 2]
```diff
@@ -930,42 +921,11 @@ proc get_ports { args } {
...
   if [info exists keys(-filter)] {
-    set ports [filter_ports1 $keys(-filter) $ports]
+    set ports [filter_objs $keys(-filter) $ports filter_ports "port" 2366]
   }
   return $ports
 }
 
-variable filter_regexp1 {@?([a-zA-Z_]+) *(==|!=|=~|!~) *([0-9a-zA-Z_\*]+)}
-variable filter_or_regexp "($filter_regexp1) *\\|\\| *($filter_regexp1)"
-variable filter_and_regexp "($filter_regexp1) *&& *($filter_regexp1)"
```
Comment (jjcherry56) [inline sdc/Sdc.tcl:940]: still need these

File: sdc/Sdc.tcl [hunk 3]
```diff
@@ -406,6 +406,36 @@ proc current_design { {design ""} } {
 
 ################################################################
 
+# Generic filter_objs command.
```
Comment (jjcherry56) [inline sdc/Sdc.tcl:409]: not a command; suggest Generic get_* filter

File: sdc/Sdc.tcl [hunk 4]
```diff
@@ -406,6 +406,36 @@ proc current_design { {design ""} } {
...
+proc filter_objs { filter objects filter_function filter_type error_code } {
```
Comment (jjcherry56) [inline sdc/Sdc.tcl:410]: prefer filter_objects

File: sdc/Sdc.i [hunk 5]
```diff
@@ -1325,13 +1325,58 @@ filter_pins(const char *property,
...
+		const char *op,
+		const char *pattern,
+		LibertyPortSeq *pins)
+{
+  return filter_objects<LibertyPort>(property, op, pattern, pins);
+}
+
+LibertyLibrarySeq
+filter_liberty_libraries(const char *property,
+	    const char *op,
```
Comment (jjcherry56) [inline sdc/Sdc.i:1357]: format

File: test/sdc_filter.tcl [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# Helper functions
+proc get_full_names { collection } {
+  set full_names {}
+  foreach name $collection {
+    lappend full_names [get_full_name $name]
+  }
+  return $full_names
+}
```
Comment (jjcherry56) [inline test/sdc_filter.tcl:8]: just use report_object_full_names from tcl/CmdUtil.tcl

File: test/sdc_filter.tcl
Comment (jjcherry56) [inline test/sdc_filter.tcl:1]: prefer get_filter.tcl to be consistent with the other private regressions (it is filtering get functions, not sdc)

**Comment by akashlevy** [comment]
All revisions made. Also added error codes into `messages.txt`, let me know if it's ok

File: tcl/Sta.tcl [hunk 7]
```diff
@@ -171,7 +171,7 @@ proc get_timing_edges_cmd { cmd cmd_args } {
     cmd_usage_error $cmd
   }
   if [info exists keys(-filter)] {
-    set arcs [filter_timing_arcs1 $keys(-filter) $arcs]
+    set arcs [filter_objs $keys(-filter) $arcs "timing arc" 541]
```
Comment (jjcherry56) [inline tcl/Sta.tcl:174]: missing filter_function arg

File: sdc/Sdc.tcl [hunk 8]
```diff
@@ -406,6 +406,36 @@ proc current_design { {design ""} } {
 
 ################################################################
 
+# Generic get_* filter.
+proc filter_objs { filter objects filter_function filter_type error_code } {
```
Comment (jjcherry56) [inline sdc/Sdc.tcl:410]: filter_type is more accurately object_type

File: test/get_filter.tcl [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+read_liberty asap7_seq.lib.gz
+read_liberty asap7_simple.lib.gz
+read_verilog reg1_asap7.v
+link_design top
+create_clock -name clk -period 500 {clk1 clk2 clk3}
+create_clock -name vclk -period 1000
+
+# Test filters for each SDC command
+puts "get_cells"
+puts [report_object_full_names [get_cells -filter liberty_cell==BUFx2_ASAP7_75t_R *]]
```
Comment (jjcherry56) [inline test/get_filter.tcl:12]: no puts required for calls to report_object_full_names

File: test/get_filter.tcl [hunk 10]
```diff
@@ -0,0 +1,31 @@
+# Read in design and libraries
+read_liberty asap7_invbuf.lib.gz
```
Comment (jjcherry56) [inline test/get_filter.tcl:2]: I pushed an abbreviated asap7 library that will read faster for reg1_asap7.v
test/asap7_small.lib.gz

**Comment by akashlevy** [comment]
I updated your recently added `etc/FindMessages.tcl` to support the `filter_objs` TCL expressions

File: etc/FindMessages.tcl [hunk 11]
```diff
@@ -13,7 +13,7 @@ exec tclsh $0 ${1+"$@"}
 # Find warning/error message IDs and detect collisions.
 # Usage: FindMessages.tcl > doc/messages.txt
 
-proc scan_file { file warn_regexp } {
+proc scan_file { file warn_regexp is_filter_objs } {
```
Comment (jjcherry56) [inline etc/FindMessages.tcl:16]: uh, no. that is too specific. I'd rather see the same error messages used by all the filter_objs callers

**Pull request merged**